File: 351003/Komasilo/task310/src/main/java/by/bsuir/task310/service/AuthorService.java
package by.bsuir.task310.service;

import by.bsuir.task310.dto.AuthorRequestTo;
import by.bsuir.task310.dto.AuthorResponseTo;
import by.bsuir.task310.exception.DuplicateException;
import by.bsuir.task310.exception.EntityNotFoundException;
import by.bsuir.task310.mapper.AuthorMapper;
import by.bsuir.task310.model.Author;
import by.bsuir.task310.repository.AuthorRepository;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.CachePut;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class AuthorService {

    private final AuthorRepository repository;
    private final AuthorMapper mapper;

    public AuthorService(AuthorRepository repository, AuthorMapper mapper) {
        this.repository = repository;
        this.mapper = mapper;
    }

    @CachePut(value = "authors", key = "#result.id")
    public AuthorResponseTo create(AuthorRequestTo requestTo) {
        if (repository.existsByLogin(requestTo.getLogin())) {
            throw new DuplicateException("Author with this login already exists");
        }

        Author author = mapper.toEntity(requestTo);
        Author saved = repository.save(author);
        return mapper.toResponseTo(saved);
    }

    public List<AuthorResponseTo> getAll() {
        return repository.findAll()
                .stream()
                .map(mapper::toResponseTo)
                .toList();
    }

    @Cacheable(value = "authors", key = "#id")
    public AuthorResponseTo getById(Long id) {
        Author author = repository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Author not found"));
        return mapper.toResponseTo(author);
    }

    @CachePut(value = "authors", key = "#requestTo.id")
    public AuthorResponseTo update(AuthorRequestTo requestTo) {
        if (!repository.existsById(requestTo.getId())) {
            throw new EntityNotFoundException("Author not found");
        }

        Author author = mapper.toEntity(requestTo);
        Author updated = repository.save(author);
        return mapper.toResponseTo(updated);
    }

    @CacheEvict(value = "authors", key = "#id")
    public void delete(Long id) {
        if (!repository.existsById(id)) {
            throw new EntityNotFoundException("Author not found");
        }

        repository.deleteById(id);
    }
}

File: 351003/Kolbeko/app/services/author_service.py
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from app.repository.db import author_repo
from app.models.author import Author
from app.schemas.author import AuthorRequestTo, AuthorResponseTo
from app.core.exceptions import AppException

class AuthorService:
    async def create(self, session: AsyncSession, dto: AuthorRequestTo) -> AuthorResponseTo:
        query = select(Author).where(Author.login == dto.login)
        result = await session.execute(query)
        existing_author = result.scalar_one_or_none()
        
        if existing_author:
            raise AppException(403, "Login already exists", 17)

        # Если логин свободен, создаем
        author = await author_repo.create(session, dto.model_dump())
        return AuthorResponseTo.model_validate(author, from_attributes=True)

    async def get_all(self, session: AsyncSession, page: int = 1, size: int = 100):
        authors = await author_repo.get_all(session, limit=size, offset=(page - 1) * size)
        return [AuthorResponseTo.model_validate(a, from_attributes=True) for a in authors]

    async def get_by_id(self, session: AsyncSession, id: int):
        res = await author_repo.get_by_id(session, id)
        if not res: raise AppException(404, "Author not found", 1)
        return AuthorResponseTo.model_validate(res, from_attributes=True)

    async def update(self, session: AsyncSession, id: int, dto: AuthorRequestTo) -> AuthorResponseTo:
        data = dto.model_dump(exclude={'id'})
        updated = await author_repo.update(session, id, data)
        if not updated: raise AppException(404, "Author not found", 2)
        return AuthorResponseTo.model_validate(updated, from_attributes=True)

    async def delete(self, session: AsyncSession, id: int):
        if not await author_repo.delete(session, id):
            raise AppException(404, "Author not found", 3)